node_modules/gh-auth/lib/api.js

```diff
@@ -87,6 +100,110 @@ var registerLocalAuthentication = function() {
 };
 
 /**
+ * Register the Passport Shibboleth authentication strategy
+ *
+ * @api private
+ */
+var registerShibbolethAuthentication = function() {
+    passport.use(new ShibbolethStrategy(function(req, app, headers, callback) {
+
+        log().debug({
+            'app': app.id,
+            'headers': headers
+        }, 'Received Shibboleth authentication callback');
+
+        var config = ConfigAPI.config(app.id);
+        var shibExternalIdAttributes = config.shibExternalIdAttributes;
+        var shibMapDisplayname = config.shibMapDisplayname;
+        var shibMapEmail = config.shibMapEmail;
+
+        // The external ID is configurable in the admin UI as a priority list (similar to
+        // how `mod_shib` works). We try to find an attribute in the released set that matches
+        // any of the configurable attributes. Rather than relying on `mod_shib`'s `remote_user`
+        // attribute, we rely on the configured list as it allows administrators to specify
+        // attributes on a per-tenant basis. We use `remote_user` as the fall back value
+        var externalId = _getBestAttributeValue(app, shibExternalIdAttributes, headers, headers['remote_user']);
+        if (!externalId) {
+            log().error({'headers': headers, 'app': app.id}, 'No suitable attribute was found for the `externalId` attribute');
+            return callback({'code': 500, 'msg': 'No suitable attribute was found for the `externalId` attribute'});
+        }
+
+        // There are a lot of SAML attributes that may indicate a user's display name. The administrator
+        // should provide a suitable priority list to construct the displayName. If no suitable value was
+        // returned from the mapping, we fall back to the `remote_user` attribute, as this is always provided
+        var displayName = _getBestAttributeValue(app, shibMapDisplayname, headers, headers['remote_user']);
+
+        // Set the optional profile parameters
+        var opts = {};
+
+        // Get an email address from the provided headers. We fall back to `eppn` which may not always
+        // contain a valid email address. We only set the `email` value if a valid email has been provided
+        var email = _getBestAttributeValue(app, shibMapEmail, headers, headers['eppn']);
+        if (email) {
+            var validator = new Validator();
+            validator.check(email, {'code': 400, 'msg': 'Invalid email'}).isEmail();
+            if (!validator.hasErrors()) {
+                opts['email'] = email;
+            } else {
+                log().warn({'email': email, 'externalId': externalId, 'app': app.id}, 'A user signed in via Shib with an invalid email address');
```

Comment:
I'm hesitant to dump potentially large objects in the logs. If we're going to use syslog in production (let's pray we don't, but suppose we do) we're limited to a certain number of characters per log statement. If you push more, syslog chops the statement up into parts it can deal with, but it loses information about where it came from. You end up with a log directory that is totally messed up. (We currently have this problem in OAE).
Sticking the headers in there would cause us to certainly cross that threshold.